compiler/AST/iterator.cpp

```diff
@@ -68,6 +69,267 @@ IteratorInfo::IteratorInfo() :
+  a group for it. We are interested in a group only when it maps
+  from a serial iterator to its parallel counterparts. So when
+  a CallExpr invokes a parallel iterator directly, iterator groups
+  are not updated.
+
+SERIAL --> PARALLEL
+
+* Each parallel iterators is attempted to be resolved by resolving
+  a "representative call". It is created as follows:
+   - It seeks a function of the same name as the serial iterator's.
+   - The actual arguments are the serial iterator's formals, plus 'tag'.
+
+* That way there is a 1:1 relationship between the serial iterator
+  its standalone counterpart; ditto leader. Because the choice
+  of the parallel iterator is affected only by the serial iterator.
+  This choice is not affected by the "original call" i.e. one that
+  led to resolving the serial iterator.
+
+* The "representative call" is placed next to the "original call".
+  To strengthen confidence in this 1:1 relationship, it seems like
+  the representative call should go next to the definition
+  of the serial iterator. However, that would cause visibility issues.
+
+LOGISTICS  
+
+* Only the availability of the standalone and leader iterators
+  is detected. Their bodies are not resolved, to avoid encountering
+  (and generating) potential compile-time errors when those iterators
+  are not actually used in the program.
+
+* The absence of a parallel iterator is not an error. An error may
+  be issued later when+if this iterator is required for a parallel
+  computation.
+
+* The follower iterator is not sought at this time.  to seek it we would
+  need to know what the leader iterator yields. For that we may have to
+  resolve the **body** of the leader. Which we are not doing here.
+
+*/
+
+std::map<FnSymbol*,IteratorGroup*> iteratorGroups;
+
+IteratorGroup::IteratorGroup() :
+  serial(NULL), standalone(NULL), leader(NULL), follower(NULL) {}
+
+static bool isGroupable(FnSymbol* it) {
+  return it->hasFlag(FLAG_ITERATOR_FN) ||
+         it->hasFlag(FLAG_FN_RETURNS_ITERATOR);
+}
+
+// Return the IteratorGroup for the iterator 'it'.
+// If none, return NULL.
+IteratorGroup* getIteratorGroup(FnSymbol* it) {
+  if (! isGroupable(it))             return NULL;
+  std::map<FnSymbol*,IteratorGroup*>::iterator itig = iteratorGroups.find(it);
+  if (itig == iteratorGroups.end())  return NULL;
+  else                               return itig->second;
+}
+
+// Implements the rules listed under "Iterator Groups" in iterator.cpp.
+void resolveAlsoParallelIterators(FnSymbol* serial, Expr* call) {
+  if (! serial || ! isGroupable(serial))
+    return;  // not an iterator
+
+  IteratorGroup*& igroupRef = iteratorGroups[serial];
+
+  if (igroupRef != NULL)
+    return;  // already taken care of
+
+  for_formals(formal, serial)
+    if (formal->name == astrTag && formal->type == gLeaderTag->type)
+      // 'serial' is actually a parallel iterator. Since we did not come
+      // from a serial iterator, we will not update iterator groups.
```

Comment:
Shouldn't this also check for a standalone iterator?
Besides that, I'd expect that we would update the iterator groups if we did start with a parallel iterator. Simpler that way.

Reply:
This checks for any formal of the leader tag's type. This includes standalone and follower.